shouldn't be using tc_dissect_pdus in dissect_frame

freehaha opened the pull request:
should just iterate through the messages instead as it's not a tcp stream anymore

freehaha committed:
parse message directly instead of calling tcp_dissect_pdus

ayyaruq commented:
I was wondering about this when I did it that way tbh. It messes up some of the protocol metadata too so I'll have a go at fixing it asap. Thanks for confirming I'm not overthinking it!

freehaha committed:
interrupt message dissecting after a broken message is found

freehaha commented:
Creating subsets should be cheap enough to do it like this. Let me know what you think.

ayyaruq commented on src/packet-ffxiv.c:
```diff
@@ -103,9 +107,9 @@ static int dissect_message(tvbuff_t *tvb, packet_info *pinfo, proto_tree *tree,
+  add_new_data_source(pinfo, tvb_new_subset_length(tvb, 0, header.length), "Message Data");
 
-  return tvb_captured_length(tvb);
+  return header.length;
```
Any reason not to assign the `tvb_new_subset_length(tvb, 0, header.length)` to a new `tvbuff_t` and then return `tvb_captured_length(new_tvb)`?

freehaha replied:
no particular reason, but since this subset is created only for adding new data source I don't see the reason to keep a pointer to it and returning header.length because that's how much we have processed. (assuming partial/malformed packets are dealt with already)

ayyaruq replied:
I was thinking more just for consistency than anything. It doesn't really matter, although I find nested methods a bit gross.

ayyaruq commented on src/packet-ffxiv.c:
```diff
@@ -145,8 +150,15 @@ static int dissect_frame(tvbuff_t *tvb, packet_info *pinfo, proto_tree *tree, vo
-  tcp_dissect_pdus(next_tvb, pinfo, frame_tree, TRUE, BLOCK_HEADER_LEN,
-    get_message_length, dissect_message, data);
+  offset = 0;
+  for(int i = 0; i < header.blocks; ++i) {
```
Not `i++`? I would maybe use a `while` loop here though since the actual variable isn't important and it will need to run at least once, and saves from having to guard for `header.blocks` being 0, although I doubt that happens.

freehaha replied:
`i++` gets optimized here to `++i` anyway? was thinking about verifying the number of blocks here hence the `for` but we can also verify it using the leftovers.

ayyaruq replied:
Relying on optimisations between 3 separate compilers can result in very unexpected behaviour. I don't mind a for loop but since it isn't used internally for iterating anything, while just seems more appropriate.

ayyaruq commented on src/packet-ffxiv.c:
```diff
@@ -145,8 +150,15 @@ static int dissect_frame(tvbuff_t *tvb, packet_info *pinfo, proto_tree *tree, vo
+  for(int i = 0; i < header.blocks; ++i) {
+    message_tvb = tvb_new_subset_remaining(next_tvb, offset);
+    length = dissect_message(message_tvb, pinfo, frame_tree, data);
+    if(length < 0) {
```
This should probably be `length < BLOCK_HEADER_LEN` since even if there is no payload in the message, the header still exists. If we have an incomplete header, something is wrong.

ayyaruq commented on src/packet-ffxiv.c:
```diff
@@ -145,8 +150,15 @@ static int dissect_frame(tvbuff_t *tvb, packet_info *pinfo, proto_tree *tree, vo
 
-  tcp_dissect_pdus(next_tvb, pinfo, frame_tree, TRUE, BLOCK_HEADER_LEN,
-    get_message_length, dissect_message, data);
+  offset = 0;
```
Offset is the offset in the current TVB and at this point should be set to the start of the payload data. Resetting it to zero would include the frame header in the message_tvb. I would probably also rename `next_tvb` to `payload_tvb` just so it's clearer what it gets used for.

freehaha replied:
will rename `next_tvb` to `payload_tvb` add a new `message_offset` I suppose.

ayyaruq replied:
Thanks, it just makes things a little clearer. We could also just use offset as the message offset since the original offset is only used for `build_frame_header` which can take it as `0` to begin with. The offset is incremented by the length of the frame header but it doesn't get used. This may just be code cleanup so feel free to ignore it, but renaming the TVB would be good.

ayyaruq reviewed (changes requested):
A few questions I have but I could be misunderstanding the intent. The only really glaring thing is whitespace though, 2 spaces rather than tabs if you don't mind. I'll try get some kind of code style and syntax validation CI setup soon.

freehaha committed:
1. fix indent
2. a few extra variable to increase clarity
3. uses while loop instead to process messages

freehaha committed:
fix typo

freehaha commented:
should be better now. let me know if there are more to change

ayyaruq commented:
Some code style/cleanliness things I want to change but for now let's merge this since it's a useful fix.
